rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/nodesource/NodeSourceWindow.java

```diff
@@ -473,10 +487,68 @@ private void addAllPluginValuesToAllFormItems(Collection<PluginDescriptor> allPl
+                                        .map(field -> plugin.getPluginName() + field.getName())
+                                        .collect(Collectors.joining(";")));
+            pluginParamOrders.put(plugin.getPluginName() + INFRASTRUCTURE_PARAM_FILE_ORDER_KEY,
+                                  plugin.getConfigurableFields()
+                                        .stream()
+                                        .filter(field -> field.isFile() || field.isCredential())
+                                        .map(field -> plugin.getPluginName() + field.getName())
+                                        .collect(Collectors.joining(";")));
+        }
         List<FormItem> formItemsForField = new LinkedList<>();
+        int currentSectionSelector = -1;
         for (PluginDescriptor.Field pluginField : pluginFields) {
+            if (pluginField.getSectionSelector() != currentSectionSelector) {
+                currentSectionSelector = pluginField.getSectionSelector();
+                // so we need to add SectionItem as a Header for the section
+                // but if selector is 0 than we dont add anything (0 is default value,
+                // so if vlaue is not set, then all values would be at the top without header)
+                if (currentSectionSelector > 0) {
+                    // if there is no description then we dont do anything
+                    if (plugin.getSectionDescriptions().containsKey(pluginField.getSectionSelector())) {
+                        final String[] ids = pluginFields.stream()
+                                                         .filter(field -> field.getSectionSelector() == pluginField.getSectionSelector())
+                                                         .map(field -> plugin.getPluginName() + field.getName())
+                                                         .toArray(String[]::new);
+                        RowSpacerItem rowSpacerItem = new RowSpacerItem(plugin.getPluginName() + "separator" +
+                                                                        currentSectionSelector);
+                        allFormItems.add(rowSpacerItem);
+                        StaticTextItem staticTextItem = new StaticTextItem(plugin.getPluginName() + "staticTextItem" +
+                                                                           currentSectionSelector,
+                                                                           plugin.getSectionDescriptions()
+                                                                                 .get(pluginField.getSectionSelector()) +
+                                                                                                   ":");
+                        staticTextItem.setTitleStyle("generalParametersStyle");
+                        allFormItems.add(staticTextItem);
+                    }
+                }
+            }
```

Comment:
if() else() -> for -> if -> if -> if => The complexity of this method is too high. Please move some codes to a private method in order to diminish the complexity. The aim is to increase readability and the maintainability.

Reply:
Mael aka Sonar is right